Feat/rpc methods WIP (#71)
kadena-io/KIPs: kip-0037.md

---
KIP: '0037'
Title: kadena_getAccount_v1 API
Author: Danillo Felixdaal, Bart Huijgen and Albert Groothedde
Status: Draft
Type: Standard
Category: Interface
Created: 2025-02-17
---

# kadena_getAccount_v1 API

## Abstract

This KIP standardizes the `kadena_getAccount_v1` method to provide the active
account known by the wallet. This allows dApps to present the user with the
currently selected account and use it for creating and signing transactions.

## Motivation

The `kadena_getAccount_v1` method provides a standardized way to retrieve a
single, active account from a Kadena wallet. This method is used when the dApp
does not need to specify which account to retrieve and relies on the wallet's
current active selection.

## Specification

### Method Definition

**Method Name:** `kadena_getAccount_v1`

#### Parameters

No parameters required. The wallet can return any account that is currently
active.

### Response

```typescript
interface AccountInfo {
  accountName: string; // The unique identifier for the account.
  networkId: string; // The unique identifier for the network for this account.

>   contract: string; // Identifier for the fungible token contract.

Reviewer: Should `contract` be optional for accounts that aren't minted yet? Or do wallets always associate accounts with a contract right away?

Also, is it possible for the same `accountName` to be present in multiple contracts? If so, should this field be a list similar to `chainAccounts`?

Reviewer 2: > Also, is it possible for the same accountName to be present in multiple contracts? If so, should this field be a list similar to chainAccounts?

The wallet only returns the ACTIVE account. Since an account is always tied to a contract, this is not a list

Reviewer: Added document suggestion related to this here: https://github.com/kadena-io/KIPs/pull/71/files#r2044838690

Resolving this conversation thread

>   guard: {

Reviewer: To get ahead of potential confusion over the definition of "guard" (https://docs.kadena.io/smart-contracts/guards) here:
It might be good to add a note somewhere in the document that we're using the term "guard" here but it focuses mainly on keyset-related guards (keyset ref, single key, multi-key, and w accounts) since that the current ownership model that wallets support.

Reviewer 2: It's any guard honestly. We should maybe extend docs to include all types of guard?

Reviewer 3: I think guard here makes sense, the other place you mentioned should be changed to refer to multiple keys indeed, which is still one guard.

Reviewer 4: @alber70g it can't be smart contract related guards since you cant add them to your wallet directly. the valid guards are `Keyset` and `Keyset ref`. the current api only support `keyset` - `{keys: string[];  pred: string}`  . its good to mention it in the KIP and make the guard type more future-proof, allowing support for Keyset Ref later without breaking the API.

Reviewer: Is the consensus to add a note in the KIP that the current API only supports `Keyset` guards at the moment?

Reviewer: Adding this rewording for y'all's consideration: https://github.com/kadena-io/KIPs/pull/71/files#r2044923640

Reviewer 3: we're renaming the `guard` field to `keyset`

    keys: string[]; // Array of public keys (secret keys are omitted for security).
    pred: string; // Predicate defining key validation (e.g., "keys-all", "keys-any").
  };

>   chainAccounts: string[]; // Array of chain IDs where this account exists

Author: @barthuijgen @alber70g, i can live with this update thoughts?

>   chainAccounts: string[]; // Array of chain IDs where this account exists

Reviewer 3: Between the two I think I like `availableChains` the best. `chainAccounts` used to represent something else before, now it indeed does not make much sense.

>   chainAccounts: string[]; // Array of chain IDs where this account exists

Reviewer 3: Me and @nillo aren't sure what to call it, @alber70g what do you prefer?

>   chainAccounts: string[]; // Array of chain IDs where this account exists

Reviewer 4: What is the benefit of having chainIds in general?
This will force wallets to query all chains, which makes the response time longer (and also more error-prone) even if the wallet stores data still need to sync data since account creation might happen outside of the wallet with transferCreate.

Also, I can imagine that wallets might just send an empty array to avoid calling all chains, which would then make the response unpredictable and untrustworthy to the end user.

the user can use `accountDiscovery` from client-utils to fetch not only chainId but also the account balance. or if we want the provider returns that we can add another message like kadena_discoverAccount_v1.

>   chainAccounts: string[]; // Array of chain IDs where this account exists

Reviewer: @javadkh2 are you proposing that the API for the `kadena_getAccount_v1` be changed so that it no longer returns the chain IDs where the account is present? Have y'all discussed this somewhere else and come to a conclusion?

>   chainAccounts: string[]; // Array of chain IDs where this account exists

Reviewer 2: @LindaOrtega I think I would stick to `chainAccounts` as that's also on GraphQL, and since dapps might use that too, it'll be more consistent (https://graph.kadena.network/graphql?query=query+accounts+%7B%0A++fungibleAccount%28accountName%3A%22k%3A554754f48b16df24b552f6832dda090642ed9658559fef9f3ee1bb4637ea7c94%22%29%7B%0A++++accountName%0A++++totalBalance%0A++++chainAccounts+%7B%0A++++++accountName%0A++++++balance%0A++++%7D%0A++%7D%0A%7D)

I believe Tatum also takes that API in their endpoints

>   chainAccounts: string[]; // Array of chain IDs where this account exists

Reviewer 4: @LindaOrtega We didn’t discuss this outside of this thread. I just thought this could add complexity to the wallet side and is also still not reliable on the dApp side. But I can see the benefit of having it if wallets provide it correctly.
For now, we can keep it, but let’s keep an eye on how wallets provide it.

> }

Author: I don't think this is a assumption that should be made, its either there or its not, if we want it to be there we should enforce it having that said. I seems like a UI thingy to me, the wallet provides the account <- the Dapp itself could provide the ability to add a label to it. Currently none of the wallets support this, so i don't think there is a necessity for this, and should leave labeling to the dapp itself. @barthuijgen @alber70g  thoughts?

> }

Reviewer 2: Chainweaver v2 and v3 both have support for aliasses though. If a user is actively using aliasses, it would be nice to have it. I would suggest not to have it nullable, but empty string if it's not there. However current APIs don't support it, correct?

> }

Reviewer: Did y'all come to a conclusion on whether to include an optional `label` field in this version of the API?

     due to configuration issues.

## Examples

### Retrieve Active Account from Wallet

#### Example Request

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "kadena_getAccount_v1"
}
```

#### Example Response

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "accountName": "k:35d381861d9a6cfb84bef47930b386d1a3b4f3742c8e87772a9e6b43c0e7b5f7",
    "networkId": "mainnet01",
    "contract": "coin",
    "guard": {
      "keys": [
        "35d381861d9a6cfb84bef47930b386d1a3b4f3742c8e87772a9e6b43c0e7b5f7"
      ],
      "pred": "keys-all"
    },
    "chainAccounts": ["0", "1"]
  }
}
```

### Error Response Example

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "error": {
    "code": -32603,
    "message": "Internal server error"
  }
}
```

## Backwards Compatibility

- **Legacy Methods:** Wallets **MAY** support older methods for retrieving
  account information alongside `kadena_getAccount_v1`. Legacy methods **MUST**
  be documented as deprecated and phased out over time.

- **Configuration Consistency:** Implementations **MUST** ensure that the
  returned account information matches the wallet's active configuration.

## Security Considerations

> - **Data Accuracy:** Ensure that the returned account information is accurate

Author: The KIP is written for wallets to implement the RPC methods, so the remarks is for the creator of the wallets, i could add, "The wallet creator implementing the RPC methods.." Ensure that the returned account information is accurate before it but that seems kind of redundant to me.

> - **Data Accuracy:** Ensure that the returned account information is accurate

Reviewer: Yes, adding it here can be redundant. But I added some rewording suggestions to the abstracts of the KIP 37 and 38 that make it a little more clear who will be implementing vs. consuming this RPC method.
- https://github.com/kadena-io/KIPs/pull/71/files#r2044896167
- https://github.com/kadena-io/KIPs/pull/71/files#r2044907844

Resolving this comment thread.

---
KIP: "0037"
Title: kadena_getAccount_v1 API
Author: Danillo Felixdaal, Bart Huijgen and Albert Groothedde
Status: Draft
Type: Standard
Category: Interface
Created: 2025-02-17
---

# kadena_getAccount_v1 API

## Abstract

> This KIP standardizes the `kadena_getAccount_v1` method to provide the active

Reviewer: ```suggestion
This KIP standardizes the `kadena_getAccount_v1` method, which wallet creators will implement to provide the active
```

> This KIP standardizes the `kadena_getAccount_v1` method to provide the active
> account known by the wallet. This allows dApps to present the user with the
> currently selected account and use it for creating and signing transactions.

Reviewer: ```suggestion
This KIP standardizes the `kadena_getAccount_v1` RPC method, which wallet creators will implement to provide the active
account known by the wallet. A dApps can use this method to present the user with the
currently selected account and use it for creating and signing transactions.
```

## Motivation

The `kadena_getAccount_v1` method provides a standardized way to retrieve a
single, active account from a Kadena wallet. This method is used when the dApp
does not need to specify which account to retrieve and relies on the wallet's
current active selection.

## Specification

### Method Definition

**Method Name:** `kadena_getAccount_v1`

#### Parameters

No parameters required. The wallet can return any account that is currently
active.

### Response

```typescript
interface AccountInfo {

>   accountName: string; // The unique identifier for the account.
>   networkId: string; // The unique identifier for the network for this account.
>   contract: string; // Identifier for the fungible token contract.

Reviewer: ```suggestion
  accountName: string; // The unique identifier for the account.
  networkId: string; // The unique identifier for the network for this account.
  contract: string; // Identifier for the fungible token contract. This is a non-empty string since wallets MUST always associate new accounts with a contract.
```

  guard: {
    keys: string[]; // Array of public keys (secret keys are omitted for security).
    pred: string; // Predicate defining key validation (e.g., "keys-all", "keys-any").

>   };

Reviewer: no longer needed with field being renamed to `keyset`

>   };

Reviewer: they renamed the field `keyset`

  chainAccounts: string[]; // Array of chain IDs where this account exists
}
```

> **Note:** The `chainAccounts` can be empty when there are no minted accounts
> on-chain.

### Error Responses

In case of failure, the wallet **MUST** return an error according to the
[JSON-RPC 2.0 specification](https://www.jsonrpc.org/specification#error_object):

```typescript
interface ErrorResponse {
  id: number;
  jsonrpc: "2.0";
  error: {
    code: number;
    message: string;
    data?: object;
  };
}
```

#### Error Types

| Code   | Message                          |
| ------ | -------------------------------- |
| -32603 | Internal server error            |
| -32008 | No account information available |

### Behavior Requirements

1. **Account Retrieval:**

>    - **MUST** return accurate and up-to-date information about the active

Reviewer 3: This spec defines only the communication protocol, these kind of things are up to the wallets to handle. If this requirement suggests otherwise I'd rather remove the requirement from the KIP.

---
KIP: '0037'
Title: kadena_getAccount_v1 API
Author: Danillo Felixdaal and Bart Huijgen
Status: Draft
Type: Standard
Category: Interface
Created: 2025-02-17
---

> # kadena_getAccount_v1 API

Reviewer 4: it seems this method has two use cases so how about creating two methods?
one for getting the active account, and another for retrieving an account by passing public key and contract. 
* kadena_getActiveAccount (no arg)
* kadena_retriveAccount (or kadena_getAccount) ({publicKey, contract})

Author: @alber70g this feedback was made on your suggestions of passing the publickey, contract as optionals in order to use them as a filter. what you think of @javadkh2 suggestion?

## Abstract

We propose the `kadena_getAccount_v1` method to provide an active account known by the wallet. This allows dApps to present the user with the selected account and allow them to use it for creating and signing transactions.

## Motivation

> The `kadena_getAccount_v1` method provides a standardized way to retrieve a list of accounts from a Kadena wallet.

Reviewer 4: does it return a list or only one?

Author: only one , the getAccount(s) returns a list

## Specification

### Method Definition

**Method Name:** `kadena_getAccount_v1`

### Request Parameters

```typescript
interface GetAccountRequest {
  account?: {
    publicKey: string
    contract: string
  }
}
```

### Response

```typescript

> interface AccountInfo {
>   accountName: string
>   fungibleContract: string
>   chainAccounts?: Array<{
>     chainId: ChainId
>     guard: {
>       keys: string[]
>       pred: string
>     }
>   }>

Reviewer 4: This type shows that an account can have a different guard on different chains, which I believe is incorrect
I believe the guard should be directly bound to the account, not the chains.
An account is a combination of an address and a guard. This approach aligns with principal accounts, 
For vanity accounts, if the guard is different, then the wallet and users should treat them as separate accounts.
```suggestion
interface AccountInfo {
  address: string
  fungibleContract: string
  guard: {
      keys: string[]
      pred: string
   }
  availableChains?: Array<{
    chainId: ChainId
  }>
```

Author: i was under the assumption that a account "CAN" have a different guard on different chains @alber70g @jermaine150

Author: @barthuijgen

Reviewer 2: I believe the most intuitive way to think about accounts is to disconnect them from the guard. A user will think the account is the same until they get the question to define which guard to use. 

Separating accounts based on guard is confusing

Author: "changed"

---
KIP: '0037'
Title: kadena_getAccount_v1 API
Author: Danillo Felixdaal and Bart Huijgen
Status: Draft
Type: Standard
Category: Interface
Created: 2025-02-17
---

# kadena_getAccount_v1 API

## Abstract

We propose the `kadena_getAccount_v1` method to provide an active account known by the wallet. This allows dApps to present the user with the selected account and allow them to use it for creating and signing transactions.

## Motivation

The `kadena_getAccount_v1` method provides a standardized way to retrieve a list of accounts from a Kadena wallet.

## Specification

### Method Definition

**Method Name:** `kadena_getAccount_v1`

### Request Parameters

```typescript
interface GetAccountRequest {
  account?: {
    publicKey: string
    contract: string
  }
}
```

### Response

```typescript
interface AccountInfo {
  accountName: string
  fungibleContract: string
  chainAccounts?: Array<{
    chainId: ChainId
    guard: {
      keys: string[]
      pred: string
    }

>   }>

Author: @barthuijgen

  "params": {
    "accounts": {
      "publicKey": "60870392901728fe6bb773534d71272d0851b587cb91cb859a303b20f1025069",
      "contract": "coin"
    }
  }
}
```

## Example response by public key

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "accountName": "w:Gkx0oHalSq0TpNQjx7Q4UTgZteaQ1QwnBH3hR60VCuY:keys-all",
    "fungibleContract": "coin",
    "chainAccounts": [
      {
        "chainId": "0",
        "guard": {
          "keys": [
            "60870392901728fe6bb773534d71272d0851b587cb91cb859a303b20f1025069",
            "49a7943c8a29962d81142067da4f4fa98940fee89efc4780b86adef9546b493a"
          ],
          "pred": "keys-all"
        }
      }
    ]
  }
}
```

## Backwards Compatibility

- Legacy Methods: Wallets MAY support older methods for retrieving account information alongside kadena_getAccounts_v1. Legacy methods MUST be documented as deprecated and phased out over time.
- Configuration Consistency: Implementations MUST ensure that the returned account information matches the wallet’s active configuration.

## Security Considerations

- Data Accuracy: Ensure that the returned account information is accurate and verified.

> - Event Notification: Emit account change events to keep dApps updated and prevent configuration mismatches.

Reviewer 2: Modify this and reference the optional KIP that describes events. In that KIP you can add that WalletConnect supports events, and the wallet-adapter will implement those events on a best-effort basis. E.g. when a wallet provider has support for event-based communication this can then propagate to the dapp, or can be emulated by the wallet-adapter itself

## Abstract

We propose the `kadena_getAccount_v1` method to provide an active account known by the wallet. This allows dApps to present the user with the selected account and allow them to use it for creating and signing transactions.
This KIP standardizes the `kadena_getAccount_v1` method to provide the active
account known by the wallet. This allows dApps to present the user with the
currently selected account and use it for creating and signing transactions.

## Motivation

The `kadena_getAccount_v1` method provides a standardized way to retrieve a list of accounts from a Kadena wallet.
The `kadena_getAccount_v1` method provides a standardized way to retrieve a
single, active account from a Kadena wallet. This method is used when the dApp
does not need to specify which account to retrieve and relies on the wallet's
current active selection.

## Specification

### Method Definition

**Method Name:** `kadena_getAccount_v1`

### Request Parameters
#### Parameters

```typescript
interface GetAccountRequest {
  account?: {
    publicKey: string

>     contract: string

Reviewer 2: Comment from @javadkh2: when the dapp knows which contract to use, they should be able to select it. It's about the right UX for the user so they don't have to think about it. It can be optional though
